src/main/java/se/liu/ida/hefquin/engine/data/utils/SolutionMappingsIteratorWithSolMapFilter.java

```diff
@@ -0,0 +1,46 @@
+package se.liu.ida.hefquin.engine.data.utils;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+
+/**
+ * This is an iterator of solution mappings that consumes another iterator
+ * and passes on only the solution mappings that are compatible to a given
+ * solution mapping.
+ */
+public class SolutionMappingsIteratorWithSolMapFilter implements Iterator<SolutionMapping>
+{
+	protected final Iterator<SolutionMapping> input;
+	protected final SolutionMapping sm;
+
+	protected SolutionMapping nextOutputElement = null;
+
+	public SolutionMappingsIteratorWithSolMapFilter( final Iterator<SolutionMapping> input, final SolutionMapping sm ) {
+		this.input = input;
+		this.sm = sm;
+	}
+
+	@Override
+	public boolean hasNext() {
+		while ( nextOutputElement == null && input.hasNext() ) {
+			final SolutionMapping nextInputElement = input.next();
+			if ( SolutionMappingUtils.compatible(sm, nextInputElement) ) {
+				nextOutputElement = nextInputElement;
+			}
+		}
+
+		return ( nextOutputElement != null );
```

Comment:
After assigning the first matching solution mapping to 'nextOutputElement', method hasNext() will always return **True** as ''nextOutputElement' is not initlized when calling hasNext() the next time.

Reply:
Good catch!

However, your proposed solution in commit https://github.com/LiUSemWeb/HeFQUIN/pull/52/commits/364e773c192d95858c4294380dc927904a80cad1 is also not correct, for the following two reasons: First, simply setting `nextOutputElement` to `null` at the begin of `hasNext()` is problematic because `hasNext()` may be called multiple times before `next()` is called; if that's the case, some output elements may never be returned. Second, removing the call of `hasNext()` at the begin of `next()` is problematic because `next()` may actually be called without calling `hasNext()`; for such cases, `hasNext()` has to be called from within `next()` to make sure we are in a reliable state.

I have another solution that I will commit.